### PR TITLE
CoverBrowser: use bookinfo cache in Classic mode

### PR DESCRIPTION
**Pull request by hius07**
Getting bookinfo from the cache db is much faster than opening a book, let's use it in Classic mode too.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/11404)

### CODE REVIEW

File: frontend/apps/filemanager/filemanagerutil.lua [hunk 1]
```diff
@@ -260,38 +260,40 @@ function filemanagerutil.genShowFolderButton(file, caller_callback, button_disab
...
 end
 
-function filemanagerutil.genBookDescriptionButton(file, caller_callback, button_disabled)
+function filemanagerutil.genBookDescriptionButton(file, bookinfo, caller_callback, button_disabled)
+    local description = bookinfo and bookinfo.description
     return {
         text = _("Book description"),
-        id = "book_description", -- used by covermenu
-        enabled = not button_disabled,
+        enabled = not (button_disabled or bookinfo) or description and true or false,
```
Comment (poire-z) [inline frontend/apps/filemanager/filemanagerutil.lua:292]: A bit hard to read this logic and guess the result of that. Could some parens help ?

I was trying to guess if all of this works when in classic mode and the coverbrowser plugin is disabled.
Can I assume all this work in that case and that these buttons are enabled, and it will go open the book to get metadata?

File: plugins/coverbrowser.koplugin/covermenu.lua [hunk 2]
```diff
@@ -232,8 +232,7 @@ function CoverMenu:updateItems(select_number)
     -- we replace it by ours.
     -- (FileManager may replace file_chooser.showFileDialog after we've been called once, so we need
     -- to replace it again if it is not ours)
-    if not self.showFileDialog_ours -- never replaced
-            or self.showFileDialog ~= self.showFileDialog_ours then -- it is no more ours
+    if self.showFileDialog and self.showFileDialog ~= self.showFileDialog_ours then
         -- We need to do it at nextTick, once FileManager has instantiated
         -- its FileChooser completely
```
Comment (poire-z) [inline plugins/coverbrowser.koplugin/covermenu.lua:237]: (Not really want to re-read all my old replacement logic) but why this change ? Simplification/optimisation (but may be a little less obvious to see the logic), or needed change because of all the other changes? Can self.showFileDialog not exist?

Comment (hius07) [reply plugins/coverbrowser.koplugin/covermenu.lua:237]: It tried to patch `showFileDialog` when showing History and Collection.

File: frontend/apps/filemanager/filemanagerutil.lua (hunk 1 above)
Comment (hius07) [reply frontend/apps/filemanager/filemanagerutil.lua:292]: Yes, if the coverbrowser is disabled, the book is opened to get the requested info.

File: plugins/coverbrowser.koplugin/covermenu.lua (hunk 2 above)
Comment (hius07) [reply plugins/coverbrowser.koplugin/covermenu.lua:237]: > Can self.showFileDialog not exist?

Yes, when showing History. And
https://github.com/koreader/koreader/blob/b832d43d985778e36841b7d08cf3db5436f26249/plugins/coverbrowser.koplugin/main.lua#L646

File: frontend/apps/filemanager/filemanagerutil.lua (hunk 1 above)
Comment (hius07) [reply frontend/apps/filemanager/filemanagerutil.lua:292]: > A bit hard to read this logic

Yes it is because the logic is tricky.
`button_disabled == true` for deleted files. But if we have (remained) description in cache, the button must be enabled. If there is bookinfo but there is no description, the button must be disabled. But if the file is supported and there is no bookinfo (yet), the button must be enabled.

File: plugins/coverbrowser.koplugin/covermenu.lua (hunk 2 above)
Comment (poire-z) [reply plugins/coverbrowser.koplugin/covermenu.lua:237]: > It tried to patch showFileDialog when showing History and Collection.

Sorry for being dense, but you are becoming as precise as @Frenzie (that is: not a lot :)) when answeing comments/questions :)

> but why this change ?

A Simplification/optimisation
B Needed change because of the other changes
C Unrelated bug you are solving

Comment (hius07) [reply plugins/coverbrowser.koplugin/covermenu.lua:237]: `CoverMenu.updateItems` is a common patch for FM, Hist and Coll.
But in Hist and Coll `showFileDialog` does not exist.

Comment (poire-z) [reply plugins/coverbrowser.koplugin/covermenu.lua:237]: Re-sorry for being extra dense.

**A B or C ?**

I'm guessing (after a few minutes reading your comments, that you could have spared me :)) it is a mix of A and C: unrelated to everything else in this PR, and we were (when in Hist and Coll) replacing (uselessly) that method that was never used? Right?

Comment (hius07) [reply plugins/coverbrowser.koplugin/covermenu.lua:237]: > we were (when in Hist and Coll) replacing (uselessly) that method that was never used? Right?

We tried to replace the method that did not exist.

File: frontend/apps/filemanager/filemanagerutil.lua (hunk 1 above)
Comment (poire-z) [reply frontend/apps/filemanager/filemanagerutil.lua:292]: Could that or bits of that be put in a comment ?

Also, can some parens help ?
At minima, instead of `this or that and true or false` (which is already hard to guess what the result is):
`(this or that) and true or false`

**Review by poire-z — APPROVED** [review]

**Comment by hius07** [comment]
I've made simple measurement (a 10,000 loop) comparing getting bookinfo from CoverBrowser db cache (reopening it on each request) and from DocSettings sdr (without opening a file).
CoverBrowser is about 3 times faster than DocSettings.

**Review by poire-z — APPROVED** [review]
